federationapi/internal/perform.go

```diff
@@ -616,6 +636,11 @@ func (r *FederationInternalAPI) PerformInvite(
 		return fmt.Errorf("gomatrixserverlib.SplitID: %w", err)
 	}
 
+	stats := r.statistics.ForServer(destination)
+	if stats.AssumedOffline() && len(stats.KnownRelayServers()) > 0 {
+		return fmt.Errorf("not performing federation since server is assumed offline with known relay servers")
```

Comment:
I abstracted all the bailouts into a common function to remove the duplication.
I'm not sure if it's a good idea to stub out anything here with a p2p version. The bailouts aren't happening just because of p2p. Doing `PerformDirectoryLookup` etc. is valid and useful in p2p, it just isn't useful to send that call to a relay server since they won't be able to provide a useful response in a span of time that would be useful.
I think `PerformInvite` would be useful to relay, but relays only handle transactions right now so I would leave that to a future feature improvement of s&f.